DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs

```diff
@@ -0,0 +1,193 @@
+        public BulkUploadResult ProcessDelegatesTable(IXLTable table, int centreId, DateTime? welcomeEmailDate = null);
+        public IXLTable OpenDelegatesTable(IFormFile file);
+    }
+
+    public class DelegateUploadFileService : IDelegateUploadFileService
+    {
+        private readonly IJobGroupsDataService jobGroupsDataService;
+        private readonly IRegistrationDataService registrationDataService;
+        private readonly IUserDataService userDataService;
+
+        public DelegateUploadFileService(
+            IJobGroupsDataService jobGroupsDataService,
+            IUserDataService userDataService,
+            IRegistrationDataService registrationDataService
+        )
+        {
+            this.userDataService = userDataService;
+            this.registrationDataService = registrationDataService;
+            this.jobGroupsDataService = jobGroupsDataService;
+        }
+
+        public BulkUploadResult ProcessDelegatesFile(IFormFile file, int centreId, DateTime? welcomeEmailDate)
+        {
+            var table = OpenDelegatesTable(file);
+            return ProcessDelegatesTable(table, centreId, welcomeEmailDate);
+        }
+
+        public BulkUploadResult ProcessDelegatesTable(IXLTable table, int centreId, DateTime? welcomeEmailDate)
+        {
+            var jobGroupIds = jobGroupsDataService.GetJobGroupsAlphabetical().Select(item => item.id).ToList();
+            var delegateRows = table.Rows().Skip(1).Select(row => new DelegateTableRow(table, row)).ToList();
+
+            foreach (var delegateRow in delegateRows)
+            {
+                if (IsDelegateRowInvalid(centreId, delegateRow, jobGroupIds, out var approved))
+                {
+                    continue;
+                }
+
+                if (approved.HasValue)
+                {
+                    UpdateDelegate(delegateRow, approved.Value, centreId);
+                }
+                else
+                {
+                    RegisterDelegate(delegateRow, welcomeEmailDate, centreId);
+                }
+            }
+
+            return new BulkUploadResult(delegateRows);
+        }
+
+        public IXLTable OpenDelegatesTable(IFormFile file)
+        {
+            var workbook = new XLWorkbook(file.OpenReadStream());
+            var worksheet = workbook.Worksheet(DelegateDownloadFileService.DelegatesSheetName);
+            var table = worksheet.Tables.Table(0);
+
+            if (!ValidateHeaders(table))
+            {
+                throw new InvalidHeadersException();
+            }
+
+            return table;
+        }
+
+        private bool IsDelegateRowInvalid(int centreId, DelegateTableRow delegateRow, IEnumerable<int> jobGroupIds, out bool? approved)
+        {
+            approved = null;
+            return !delegateRow.IsValid(jobGroupIds) || !TryGetExistingApprovedStatus(delegateRow, centreId, out approved);
+        }

```

Comment:
I feel like this method is doing a bit too much as we're doing the validation and then getting the approved status.